project_material/models/project_task_material.py

```diff
@@ -0,0 +1,317 @@
+        """Adjust procurements when modifying the quantity on a material line.
+
+        The sudo is required, because project users do not have access to stock objects.
+        """
+        if 'product_id' in vals or 'task_id' in vals:
+            for line in self:
+                line.sudo()._check_can_change_product_or_task()
+                line.sudo()._cancel_procurements()
+
+        super().write(vals)
+
+        if 'product_id' in vals or 'task_id' in vals or 'initial_qty' in vals:
+            lines_with_procurement = self.filtered(lambda l: l._should_generate_procurement())
+            for line in lines_with_procurement:
+                line.sudo()._run_procurements()
+
+        return True
+
+    @api.multi
+    def unlink(self):
+        """Cancel procurements when deleting material line.
+
+        The sudo is required, because project users do not have access to stock objects.
+        """
+        for line in self:
+            line.sudo()._check_line_can_be_deleted()
+            line.sudo()._cancel_procurements()
+        return super().unlink()
+
+    def _should_generate_procurement(self):
+        return True
+
+    def _cancel_procurements(self):
+        self.initial_qty = 0
+        self._run_procurements()
+        self.move_ids.write({'material_line_id': False})
+
+    def _run_procurements(self):
+        """Generate procurements for the material line.
+
+        Generate any required stock move (using procurements).
+
+        Reduce the quantity on existing stock moves if it exceeds the initial quantity
+        on the material line.
+        """
+        self._check_date_planned()
+        self._check_initial_qty_greater_than_zero()
+
+        precision = self._get_uom_precision()
+
+        move_qty = self._get_consumption_stock_move_qty()
+        more_move_qty_than_required = (
+            float_compare(move_qty, self.initial_qty, precision_digits=precision) > 0
+        )
+
+        if more_move_qty_than_required:
+            self._check_quantity_can_be_reduced()
+
+        self._update_procurement_quantities()
+
+        if more_move_qty_than_required:
+            self._cancel_moves_with_zero_quantity()
+
+    def _check_date_planned(self):
+        if not self.task_id.date_planned:
+            raise ValidationError(_(
+                'Before adding material to the task, you must set a planned date.'
+            ))
+
+    def _check_initial_qty_greater_than_zero(self):
+        precision = self._get_uom_precision()
+        if float_compare(self.initial_qty, 0, precision_digits=precision) < 0:
+            raise ValidationError(_(
+                'Material consumption lines can not have an initial quantity below zero. '
+                'The line {line} has a quantity of {qty}.'
+            ).format(line=self.product_id.display_name, qty=self.initial_qty))
+
+    def _update_procurement_quantities(self):
+        """Launch a procurement for the missing quantity on stock moves.
```

Comment:
Codacy found an issue: [Multi-line docstring summary should start at the second line (D213)](https://app.codacy.com/gh/Numigi/odoo-project-addons/pullRequest?prid=6119862)